src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/VsSolutionUpdatesProjectSnapshotChangeTrigger.cs

```diff
@@ -144,16 +144,26 @@ internal Task OnProjectBuiltAsync(IVsHierarchy projectHierarchy, CancellationTok
         var projectFilePath = _projectService.GetProjectPath(projectHierarchy);
         return _projectSnapshotManagerDispatcher.RunOnDispatcherThreadAsync(() =>
         {
-            var projectSnapshot = _projectManager?.GetLoadedProject(projectFilePath);
-            if (projectSnapshot is not null)
+            if (_projectManager is null)
             {
-                var workspaceProject = _projectManager?.Workspace.CurrentSolution.Projects.FirstOrDefault(
-                    wp => FilePathComparer.Instance.Equals(wp.FilePath, projectSnapshot.FilePath));
-                if (workspaceProject is not null)
+                return;
+            }
+
+            var projectKeys = _projectManager.GetAllProjectKeys(projectFilePath);
+            foreach (var projectKey in projectKeys)
+            {
+                var projectSnapshot = _projectManager?.GetLoadedProject(projectKey);
+                if (projectSnapshot is not null)
                 {
-                    // Trigger a tag helper update by forcing the project manager to see the workspace Project
-                    // from the current solution.
-                    _workspaceStateGenerator.Update(workspaceProject, projectSnapshot, cancellationToken);
+                    // TODO: Find the right project... somehow
```

Comment:
Out of curiosity what are you planning for the long-term fix here? It actually is still kind of unclear to me why the current code below is problematic - is it because of the file path comparison?

Reply:
The problem isn't the file path comparison, its the fact that its using the file path at all. For example, if you have a project file that targets `net6.0` and `net7.0`, they could have different tag helpers available, because they'll reference different versions of Asp.Net. This method gets told "Project1.csproj was built", so we look through all of _our_ projects that could possibly represent that, which in future will be two different entries, one for net6 and one for net7. We then have to go find the Roslyn project for each of those, but we have no way of mapping from our ProjectKey to Roslyn's ProjectId, so we have to find a way to work it out. The current code is just using "Project1.csproj", so for both of our representations will find the same Roslyn project, and trigger it twice, which is wrong.

This isn't a problem in this PR because we don't have multiple representations for each project file yet.

Also the solution ended up being pretty easy, because we can map trivially from a Roslyn `Project` to our `ProjectKey`.

Reply:
Makes sense! Thanks for the explanation.